### PR TITLE
Query: Relational: Fix #4412 - Ensure discriminator filtering is app…

### LINKED ISSUE #4412: TPH - "DbContext.Xs.Select(e => ...)" select all rows (=ignore discriminator)
Hi,

``` C#
public class EntityBase
{
  public int Id { get; set; }
}

public class EntityA : EntityBase
{
  public string NameA { get; set; }
}

public class EntityB : EntityBase
{
  public string NameB { get; set; }
}
```
This working great (select contains WHERE Discriminator = 'EntityA'):
``` C#
var array = context.EntityA.ToArray();
```
!!! BUT this select everything without WHERE Discriminator:
``` C#
var array = context.EntityA.Select(e => new object[] { e.NameA }).ToArray();
```

Thx for check and fix :)

### PR DESCRIPTION
**Pull request by anpete**
…lied for projection queries against inherited entities.

### CODE REVIEW

**Comment by anpete** [comment]
@maumar

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs [hunk 1]
```diff
@@ -228,12 +228,67 @@ var materializer
...
             {
+                DiscriminateProjectionQuery(entityType, selectExpression, _querySource);
+
                 shaper = new ValueBufferShaper(_querySource);
             }
 
             return shaper;
         }
 
+        private void DiscriminateProjectionQuery(
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalEntityQueryableExpressionVisitor.cs]: AddDiscriminatorPredicate? (policheck)

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs [hunk 2]
```diff
@@ -841,8 +841,9 @@ public override void Join_navigation_translated_to_subquery_composite_key()
 INNER JOIN [CogTag] AS [t] ON [g].[FullName] = (
     SELECT TOP(1) [subQuery0].[FullName]
     FROM [Gear] AS [subQuery0]
-    WHERE ([subQuery0].[Nickname] = [t].[GearNickName]) AND ([subQuery0].[SquadId] = [t].[GearSquadId])
-)",
+    WHERE (([subQuery0].[Discriminator] = N'Officer') OR ([subQuery0].[Discriminator] = N'Gear')) AND (([subQuery0].[Nickname] = [t].[GearNickName]) AND ([subQuery0].[SquadId] = [t].[GearSquadId]))
```
Comment (maumar) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs]: why is this applied here?

Comment (anpete) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/GearsOfWarQuerySqlServerTest.cs]: Inheritance.

**Comment by maumar** [comment]
:shipit:

**Pull request merged**